decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/application/internal/IntMessageService.java

```diff
@@ -39,53 +43,46 @@
+    @Qualifier("secureWebClient")
+    public WebClient.Builder webClientBuilder;
+
     @Value("${recipient.port}")
     private String PORT;
 
     /**
-     * Sends an ISO 20022 message using HTTPS to its dedicated endpoint at the recipent host
+     * Sends an ISO 20022 message using HTTPS to its dedicated endpoint at the
+     * recipent host and returns the response returned, regardless of its status
      * 
      * @param mx The ISO 20022 message to send
      * @return The HTTP response sent by the recipent host
      * @throws Exception If sending the message failed
      */
-    public ResponseEntity send(AbstractMX mx) throws Exception {
-        String host = ((BusinessAppHdrV02)mx.getAppHdr()).getTo().getFIId().getFinInstnId().getNm();
-        byte[] encoded = mx.message().getBytes("utf-8");
-
-        URL url = new URL("https://" + host + ":" + PORT + endpointOf(mx));
-
-        HttpsURLConnection conn = (HttpsURLConnection) url.openConnection();
-        conn.setRequestMethod("POST");
-        configureTruststore(conn, TRUST_STORE, TRUST_PASS);
-        conn.setDoOutput(true);
-        conn.setRequestProperty("Content-Length", "" + encoded.length);
-        conn.getOutputStream().write(encoded);
-
-        return getHttpsResponse(conn);
-    }
-
-    private ResponseEntity getHttpsResponse(HttpsURLConnection con) throws IOException {
-        InputStream reader = hasErrorResponse(con) ? con.getErrorStream() : con.getInputStream();
-        BufferedReader in = new BufferedReader(new InputStreamReader(reader, "utf-8"));
-        StringBuilder respReader = new StringBuilder();
-        String row;
-        while ((row = in.readLine()) != null) {
-            respReader.append(row + "\n");
+    public ResponseEntity<String> send(AbstractMX mx) throws Throwable {
+        String host;
+        try {
+            host = ((BusinessAppHdrV02) mx.getAppHdr()).getTo().getFIId().getFinInstnId().getNm();
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Failed to retrieve the recipient from the message.");
         }
-        HttpHeaders headers = new HttpHeaders();
-        Map<String, List<String>> conHeaders = con.getHeaderFields();
-        for (String header : conHeaders.keySet()) {
-            if (header != null) {
-                headers.add(header, conHeaders.get(header).get(0));
-            }
+        URI uri = new URI("https://" + host + ":" + PORT + endpointOf(mx));
+
+        try {
+            return webClientBuilder
+                .defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_XML_VALUE).build().post().uri(uri)
+                .bodyValue(mx.message()).retrieve().onStatus(HttpStatus::isError, (it -> {
+                    if (it.statusCode().is4xxClientError()) {
+                        // If the response is a 4xx error, it means the internal component mistakenly
+                        // validated the message. (Or the external component mistakenly flagged it)

```

Comment:
How can the internal component validate the message when it sends it to the external component, not quite sure as to how the internal component can mistakenly validate a message here?
And how can the component flag a message by mistake?

Reply:
Due to a bug in our code, or if the other financial institutions is using a custom version of our system that has a bug, which means this error would be quite severe.

Reply:
Basically, if a 4xx response is returned, the internal component thought the message was valid, but the external component did not. This should never happen if both components uses the same method to validate the messages. But it can happen due to what I stated above.

Reply:
I guess it could also happen due to a 404, but that would mean that the other financial institution uses an unsupported version of the API (or none at all)

Reply:
Shouldn't this happen if one of the parties simply doesn't trust the other? (Doesn't have the other party in their trust store)

Reply:
In that case, there wouldn't be an HTTP status as all, as a TLS alert will occur and the connection will be closed.

Reply:
But now that you mention it, it would be nice if a 403 could be returned if the client does not have authentication.

Reply:
Yeah, I think this is all right then. 👍